Engine/src/Engine/Serializer.cpp

```diff
@@ -0,0 +1,181 @@
+	}
+
+	void to_json(nlohmann::json& j, const vec3& vec)
+	{
+		j = { { "x", vec.x }, { "y", vec.y }, { "z", vec.z } };
+	}
+
+	void from_json(const nlohmann::json& j, vec3& vec)
+	{
+		vec.x = j.at("x").get<float>();
+		vec.y = j.at("y").get<float>();
+		vec.z = j.at("z").get<float>();
+	}
+
+	void to_json(nlohmann::json& j, const vec4& vec)
+	{
+		j = { { "r", vec.r }, { "g", vec.g }, { "b", vec.b }, { "a", vec.a } };
+	}
+
+	void from_json(const nlohmann::json& j, vec4& vec)
+	{
+		vec.r = j.at("r").get<float>();
+		vec.g = j.at("g").get<float>();
+		vec.b = j.at("b").get<float>();
+		vec.a = j.at("a").get<float>();
+	}
+
+	//setup only for 1 parameter matrices.
+	void to_json(nlohmann::json& j, const mat4& mat)
+	{
+		j = { { "matrix", mat[0][0]} };
+	}
+
+	void from_json(const nlohmann::json& j, mat4& mat)
+	{
+		mat = mat4(j.at("matrix").get<float>());
+	}
+}
+
+namespace Engine
+{
+	bool Serializer::tryDeserializeScene(Scene& out, const std::string& sceneFile)
+	{
+		if (!out.initializeGL())
+		{
+			GE_CORE_FATAL("Failed to initialize opengl");
+			return false;
+		}
+
+		auto data = ResourceManager::getInstance()->getJsonData(sceneFile);
+		if (!(data.find("scene") != data.end()))
+		{
+			GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+			GE_CORE_FATAL("No scene found.");
+			return false;
+		}
+
+		if (!(data["scene"].find("name") != data["scene"].end()))
+		{
+			GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+			GE_CORE_FATAL("Scene does not contain the \"name\" attribute.");
+			return false;
+		}
+
+		out.m_name = data["scene"]["name"];
+
+		//Loop through all entities and deserialize each
+		for (const auto& item : data["scene"]["entities"])
+		{
+			if (!(item.find("tag") != item.end()))
+			{
+				GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+				GE_CORE_FATAL("An entity does not contain the \"tag\" attribute");
+				return false;
+			}
+
+			Entity entity = out.createEntity(item["tag"]);
+			if (!tryDeserializeEntity(entity, item, out))
+			{
+				GE_CORE_FATAL("Unable to deserialize scene {0}", sceneFile);
+				GE_CORE_FATAL("The entity {0} has failed to serialize", item["tag"]);
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	void Serializer::serializeScene(const Scene& scene, const std::string& sceneFile)
+	{
+		//TODO: Implement serialization once we have more GUI stuff
+	}
+
+	bool Serializer::tryDeserializeEntity(Entity& out, const nlohmann::json& entity, Scene& scene)
+	{
+		//Loop through all components and deserialize each
+		for (const auto& component : entity["components"])
+		{
+			if (!(component.find("name") != component.end()))
+			{
+				GE_CORE_FATAL("Unable to deserialize entity {0}", entity["tag"]);
+				GE_CORE_FATAL("A component does not contain the \"name\" attribute");
+				return false;
+			}
+
+			auto name = component["name"].get<std::string>();
+
+			//check each component type individually
+			if (name == "CameraComponent")
```

Comment:
This can be changed to a switch statement for performance. Since c++ doesn't allow switch statements with strings, we could use a map with an enum to replicate the same behaviour.